numpy/lib/shape_base.py

```diff
@@ -89,9 +90,20 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
     outshape = asarray(arr.shape).take(indlist)
     i.put(indlist, ind)
     res = func1d(arr[tuple(i.tolist())], *args, **kwargs)
-    #  if res is a number, then we have a smaller output array
+
+    # see gh-8363
+    _strcase = False
+    _dtypes = set()
+
+    # If res is a number, then we have a smaller output array.
     if isscalar(res):
```

Comment:
What if instead of checking `isscalar` (which looks for numpy scalars), we cast `res` to a numpy array using `np.asarray`, and then check for `res.ndim == 0` on this line?

That would let us avoid making all these special cases for str dtypes -- instead, you could simply return an scalar array with dtype=object if you want the appropriate handling for strings.